### PR TITLE
[GDN] Add bounded GDN2 gate path

### PR DESCRIPTION
**Pull request by n1o**
## Summary

This PR adds an opt-in bounded gate path for `GatedDeltaNet2` and stabilizes the shared KDA lower-bound gate backward for saturated gate inputs.

The default GDN2 behavior is unchanged. When `safe_gate=True`, the layer passes the raw decay-gate preactivation into the existing in-kernel lower-bound gate and constrains the activated log-decay to `[gate_lower_bound, 0)`, with `gate_lower_bound=-7.0` by default. This permits very strong forgetting (`exp(-7) ~= 0.00091`) while avoiding numerically meaningless extreme log-decay values.

The KDA gate backward now avoids saturated `0 * inf -> nan` cases when `A_log` is very large and the lower-bound sigmoid is already saturated. It also masks inactive Triton block lanes before the `A_log` reduction so partial blocks cannot introduce NaN gradients from synthetic masked values. This matters because GDN2 reuses the KDA gate activation path when `use_gate_in_kernel=True`.

This failure mode showed up while training a model with `GatedDeltaNet2` and a MuonW-style optimizer: once the learned gate state becomes extreme, the vanilla exponential parameterization can produce non-finite forward/backward values. The fix is intentionally opt-in and does not claim that every optimizer will naturally reach that state.

The branch also includes a small model-test helper update so model tests that require optional `flash-attn` skip when that package is absent, instead of failing unrelated CI shards after the shared cache helper is touched.

Co-authored with Codex GPT-5.5 xhigh assistance.

## Test plan

Local checks:

- `python -m py_compile fla/models/utils.py fla/layers/gdn2.py fla/ops/gdn2/chunk.py fla/ops/gdn2/fused_recurrent.py fla/ops/kda/chunk.py fla/ops/kda/gate.py tests/models/test_cache.py tests/models/test_modeling_base.py tests/models/test_modeling_utils.py tests/ops/test_gdn2.py tests/ops/test_kda.py`
- `python scripts/check_header.py --check`
- `git diff --check`
- `rg -n "do_scale|hyperball|reproduce_gdn2_muon_overflow|model_change_suggestions" fla tests scripts .github README.md CONTRIBUTING.md` returned no matches

CUDA validation environment:

- GPU: NVIDIA A100-PCIE-40GB
- CUDA runtime: 13.1.1
- Python: 3.11.15
- PyTorch: 2.10.0+cu130
- Triton: 3.6.0
- Pytest: 9.1.1

CUDA tests run in the PR container on GPU 3:

- `python -m pytest -p no:cacheprovider -q tests/ops/test_kda.py::test_lowerbound_gate_extreme_a_log_saturated_backward tests/ops/test_kda.py::test_gate tests/ops/test_gdn2.py::test_layer_safe_gate tests/ops/test_gdn2.py::test_layer_safe_gate_fused_recurrent`
- Result after the partial-block mask fix: `20 passed, 14 warnings in 101.01s`

Additional CI-failure reproduction run in the same container:

- `python -m pytest -q tests/models/test_cache.py::test_fla_layer_reports_dynamic_max_length "tests/models/test_modeling_kda.py::test_generation[L2-B4-T2000-H8-D64-torch.float16]"`
- Result: `2 passed, 14 warnings in 214.01s`

Optional dependency model-test skip verification in the same container:

- `python -m pytest -q --exitfirst tests/models/test_modeling_bitnet.py tests/models/test_modeling_deltaformer.py tests/models/test_modeling_mla.py tests/models/test_modeling_moba.py tests/models/test_modeling_nsa.py tests/models/test_modeling_transformer.py`
- Result: `38 skipped, 14 warnings in 11.89s` when `flash-attn` is absent

Additional direct probe that failed before the KDA backward mask fix:

- `A_log=100`, `g=32`, `lower_bound=-7`, partial Triton block shape `B=1, T=4, H=2, D=8`
- Output after fix: `y finite True -7.0 -7.0`
- Gradients after fix: `grads finite True True True`, samples `-0.0 0.0 0.0`

Manual forced-state checks:

- Vanilla gate with `A_log=80`, `gate_preact=32` produces `loss=inf`, `max_grad=inf`
- Bounded gate with the same forced state remains finite
- Adam shows the same behavior once the forced bad gate state is already present, so the issue is not MuonW-specific at the kernel level

## Benchmark / NCU

Not run. This is a correctness/numerical robustness change. The new GDN2 path is opt-in; existing default behavior is unchanged.

## Breaking changes

No API break. `GatedDeltaNet2` gains optional `safe_gate` and `gate_lower_bound` arguments. Both are disabled by default for backward compatibility.

### CODE REVIEW

**Comment by n1o** [comment]
Updated this PR to use a narrower and more defensible fix.

The old A_log-floor / broad finite-guard approach has been replaced with an opt-in bounded GDN2 gate path that reuses the existing KDA lower-bound gate contract. I also fixed the shared KDA lower-bound gate backward so saturated gates with very large `A_log` do not produce `0 * inf -> nan` gradients.

What this PR now claims:
- default GDN2 behavior is unchanged;
- `safe_gate=True` bounds the activated GDN2 log-decay to `[gate_lower_bound, 0)`, defaulting to `[-5, 0)`;
- saturated lower-bound gate backward remains finite even for extreme `A_log` values;
- the included reproducer is a forced gate-state stress test, not proof that any optimizer naturally reaches that state.

Validation on an A100:
- `tests/ops/test_kda.py::test_lowerbound_gate_extreme_a_log_saturated_backward`
- `tests/ops/test_kda.py::test_gate`
- `tests/ops/test_gdn2.py::test_layer_safe_gate`
- `tests/ops/test_gdn2.py::test_layer_safe_gate_fused_recurrent`

Result: `18 passed, 14 warnings in 5.08s`.

**Comment by n1o** [comment]
Context for why I looked at this path: I hit the issue while training a model that uses `GatedDeltaNet2` with a MuonW-style optimizer setup. The PR does not claim the included stress script reproduces the full training run, but that training failure is what motivated checking whether the GDN2 gate path can enter numerically unsafe regimes and whether the bounded gate path avoids that class of failure.

**Comment by n1o** [comment]
Co-authorship note: this patch was developed collaboratively with Codex GPT-5.5 xhigh assistance, including the bounded-gate implementation, review iterations, and validation runs.
